drivers/i2s/i2s_nrfx_tdm.c

```diff
@@ -0,0 +1,1089 @@
+				/* Supply the same buffer as last time; it will
+				 * not be transferred anyway, as the transfer
+				 * will be stopped earlier.
+				 */
+				next.p_tx_buffer = drv_data->last_tx_buffer;
+				next.p_tx_mem_slab = drv_data->last_tx_mem_slab;
+				next.buffer_size = 1;
+			} else {
+				/* Next TX buffer cannot be supplied now.
+				 * Defer it to when the user writes more data.
+				 */
+				drv_data->next_tx_buffer_needed = true;
+				return;
+			}
+		}
+		(void)supply_next_buffers(drv_data, &next);
+	}
+}
+
+static const struct i2s_driver_api tdm_nrf_drv_api = {
+	.configure = tdm_nrfx_configure,
+	.config_get = tdm_nrfx_config_get,
+	.read = tdm_nrfx_read,
+	.write = tdm_nrfx_write,
+	.trigger = tdm_nrfx_trigger,
+};
+
+#define TDM(idx)         DT_NODELABEL(tdm##idx)
+#define TDM_CLK_SRC(idx) DT_STRING_TOKEN(TDM(idx), clock_source)
+
+#define TDM_NRFX_DEVICE(idx)                                                                       \
+	static tdm_ctrl_t tdm##idx##data;                                                          \
+	static struct tdm_buf tx_msgs##idx[CONFIG_TDM_NRFX_TX_BLOCK_COUNT];                        \
+	static struct tdm_buf rx_msgs##idx[CONFIG_TDM_NRFX_RX_BLOCK_COUNT];                        \
+	static void tdm_##idx##_irq_handler(const struct device *dev)                              \
+	{                                                                                          \
+		tdm_irq_handler(dev);                                                              \
+	}                                                                                          \
+	static void tdm_##idx##data_handler(tdm_buffers_t const *p_released, uint32_t status)      \
+	{                                                                                          \
+		data_handler(DEVICE_DT_GET(TDM(idx)), p_released, status);                         \
+	}                                                                                          \
+	PINCTRL_DT_DEFINE(TDM(idx));                                                               \
+	static const struct tdm_drv_cfg tdm_nrfx_cfg##idx = {                                      \
+		.data_handler = tdm_##idx##data_handler,                                           \
+		.pcfg = PINCTRL_DT_DEV_CONFIG_GET(TDM(idx)),                                       \
+		.clk_src = TDM_CLK_SRC(idx),                                                       \
+		.mck_frequency = DT_PROP_OR(TDM(idx), mck_frequency, 0),                           \
+	};                                                                                         \
+	static struct tdm_drv_data tdm_nrfx_data##idx = {                                          \
+		.state = I2S_STATE_READY,                                                          \
+		.p_reg = NRF_TDM##idx,                                                             \
+		.control_data = &tdm##idx##data,                                                   \
```

Comment:
ideally this data structure should not require initialization (that saves code size, as whole copy of that structure must be in rom even when single field is statically initialized). Maybe a pointer to a control_data could be moved to configuration structure too?